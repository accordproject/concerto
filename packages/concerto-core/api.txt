class AstModelManager extends BaseModelManager {
   + void constructor(object?) 
}
class BaseModelManager {
   + void constructor(object?,boolean?,Object?,boolean?,boolean?,boolean?,boolean?,processFile?) 
   + boolean isModelManager() 
   + boolean isStrict() 
   + boolean isAliasedTypeEnabled() 
   + Object accept(Object,Object) 
   + void validateModelFile(string|ModelFile,string?) throws IllegalModelException
   + Object addModelFile(ModelFile,string?,string?,boolean?) throws IllegalModelException
   + ModelFile addModel(modelInput,string?,string?,boolean?) throws IllegalModelException
   + Object updateModelFile(string|ModelFile,string?,boolean?) throws IllegalModelException
   + void deleteModelFile(string) 
   + Object[] addModelFiles(|,string[],boolean?) 
   + void validateModelFiles() 
   + Promise updateExternalModels(Object?,FileDownloader?) throws IllegalModelException
   + void writeModelsToFileSystem(string,Object?,boolean) 
   + Object[] getModels(Object?,boolean) 
   + void clearModelFiles() 
   + ModelFile getModelFile(string) 
   + string[] getNamespaces() 
   + ClassDeclaration getType(string) throws TypeNotFoundException
   + AssetDeclaration[] getAssetDeclarations() 
   + TransactionDeclaration[] getTransactionDeclarations() 
   + EventDeclaration[] getEventDeclarations() 
   + ParticipantDeclaration[] getParticipantDeclarations() 
   + MapDeclaration[] getMapDeclarations() 
   + EnumDeclaration[] getEnumDeclarations() 
   + ConceptDeclaration[] getConceptDeclarations() 
   + Factory getFactory() 
   + Serializer getSerializer() 
   + DecoratorFactory[] getDecoratorFactories() 
   + void addDecoratorFactory(DecoratorFactory) 
   + boolean derivesFrom(string,string) 
   + object resolveMetaModel(object) 
   + void fromAst(ast) 
   + void getAst(boolean?) 
   + BaseModelManager filter(FilterFunction) 
}
class Concerto {
   + void constructor(ModelManager) 
   + void validate(obj,options?) throws Error
   + ModelManager getModelManager() 
   + boolean isObject(obj) 
   + void getTypeDeclaration(obj) 
   + string getIdentifier(obj) 
   + boolean isIdentifiable(obj) 
   + boolean isRelationship(obj) 
   + void setIdentifier(obj,string) 
   + string getFullyQualifiedIdentifier(obj) 
   + string toURI(obj) 
   + void fromURI(string) throws Error
   + string getType(obj) 
   + string getNamespace(obj) 
}
   + object setCurrentTime() 
class DecoratorManager {
   + ModelManager validate(decoratorCommandSet,ModelFile[]) throws Error
   + ModelManager decorateModels(ModelManager,decoratorCommandSet,object?,boolean?,boolean?,boolean?,boolean?) 
   + ExtractDecoratorsResult extractDecorators(ModelManager,object,boolean,string) 
   + ExtractDecoratorsResult extractVocabularies(ModelManager,object,boolean,string) 
   + ExtractDecoratorsResult extractNonVocabDecorators(ModelManager,object,boolean,string) 
   + void validateCommand(ModelManager,command) 
   + Boolean falsyOrEqual(string||,string[]) 
   + void applyDecorator(decorated,string,newDecorator) 
   + void executeCommand(string,declaration,command,property?,object?,boolean?) 
   + void executePropertyCommand(property,command) 
}
   + string[] intersect() 
   + boolean isUnversionedNamespaceEqual() 
class Factory {
   + string newId() 
   + void constructor(ModelManager) 
   + Resource newResource(String,String,String?,Object?,boolean?,String?,boolean?) throws TypeNotFoundException
   + Resource newConcept(String,String,String?,Object?,boolean?,String?,boolean?) throws TypeNotFoundException
   + Relationship newRelationship(String,String,String) throws TypeNotFoundException
   + Resource newTransaction(String,String,String?,Object?,String?,boolean?) 
   + Resource newEvent(String,String,String?,Object?,String?,boolean?) 
}
class AssetDeclaration extends IdentifiedDeclaration {
   + void constructor(ModelFile,Object) throws IllegalModelException
   + string declarationKind() 
}
class ClassDeclaration extends Declaration {
   + ClassDeclaration _resolveSuperType() 
   ~ void validate() throws IllegalModelException
   + boolean isAbstract() 
   + Boolean isIdentified() 
   + Boolean isSystemIdentified() 
   + Boolean isExplicitlyIdentified() 
   + string getIdentifierFieldName() 
   + Property getOwnProperty(string) 
   + Property[] getOwnProperties() 
   + string getSuperType() 
   + ClassDeclaration getSuperTypeDeclaration() 
   + ClassDeclaration[] getAssignableClassDeclarations() 
   + ClassDeclaration[] getDirectSubclasses() 
   + ClassDeclaration[] getAllSuperTypeDeclarations() 
   + Property getProperty(string) 
   + Property[] getProperties() 
   + Property getNestedProperty(string) throws IllegalModelException
   + String toString() 
   + boolean isAsset() 
   + boolean isParticipant() 
   + boolean isTransaction() 
   + boolean isEvent() 
   + boolean isConcept() 
   + boolean isEnum() 
   + boolean isMapDeclaration() 
   + boolean isClassDeclaration() 
}
class ConceptDeclaration extends ClassDeclaration {
   + void constructor(ModelFile,Object) throws IllegalModelException
   + string declarationKind() 
}
class Declaration extends Decorated {
   + void constructor(ModelFile,Object) throws IllegalModelException
   + ModelFile getModelFile() 
   + string getName() 
   + string getNamespace() 
   + string getFullyQualifiedName() 
   + Boolean isIdentified() 
   + Boolean isSystemIdentified() 
   + string getIdentifierFieldName() 
   + string getType() 
   + String toString() 
   + boolean isEnum() 
   + boolean isClassDeclaration() 
   + boolean isScalarDeclaration() 
   + boolean isMapDeclaration() 
}
class Decorator {
   + void constructor(ClassDeclaration|Property,Object) throws IllegalModelException
   + Object accept(Object,Object) 
   + void getParent() 
   + string getName() 
   + object[] getArguments() 
   + boolean isDecorator() 
}
class DecoratorFactory {
   + Decorator newDecorator(ClassDeclaration|Property,Object) 
}
class EnumDeclaration extends ClassDeclaration {
   + void constructor(ModelFile,Object) throws IllegalModelException
   + String toString() 
   + string declarationKind() 
}
class EnumValueDeclaration extends Property {
   + void constructor(ClassDeclaration,Object) throws IllegalModelException
   + boolean isEnumValue() 
}
class EventDeclaration extends IdentifiedDeclaration {
   + void constructor(ModelFile,Object) throws IllegalModelException
   + string declarationKind() 
}
class IdentifiedDeclaration extends ClassDeclaration {
   + void constructor(ModelFile,Object) throws IllegalModelException
}
class IllegalModelException extends BaseFileException {
   + void constructor(string,ModelFile?,Object?,number,number,number,number,string?) 
}
class Introspector {
   + void constructor(ModelManager) 
   + Object accept(Object,Object) 
   + ClassDeclaration[] getClassDeclarations() 
   + ClassDeclaration getClassDeclaration(String) throws Error
}
class MapDeclaration extends Declaration {
   + void constructor(ModelFile,Object) throws IllegalModelException
   ~ void validate() throws IllegalModelException
   + string getFullyQualifiedName() 
   + ModelFile getModelFile() 
   + string getName() 
   + MapKeyType getKey() 
   + MapValueType getValue() 
   + String toString() 
   + string declarationKind() 
   + boolean isMapDeclaration() 
}
class MapKeyType extends Decorated {
   + void constructor(MapDeclaration,Object,ModelFile) throws IllegalModelException
   ~ void validate() throws IllegalModelException
   + ModelFile getModelFile() 
   + MapDeclaration getParent() 
   + string getType() 
   + String toString() 
   + boolean isKey() 
   + boolean isValue() 
   + string getNamespace() 
}
class MapValueType extends Decorated {
   + void constructor(MapDeclaration,Object) throws IllegalModelException
   ~ void validate() throws IllegalModelException
   + ModelFile getModelFile() 
   + MapDeclaration getParent() 
   + string getType() 
   + String toString() 
   + boolean isKey() 
   + boolean isValue() 
   + string getNamespace() 
}
   + ModelManager newMetaModelManager() 
   + object validateMetaModel() 
   + ModelManager modelManagerFromMetaModel() 
class ModelFile extends Decorated {
   + void constructor(ModelManager,object,string?,string?) throws IllegalModelException
   ~ ModelFile getModelFile() 
   + boolean isModelFile() 
   + string getVersion() 
   + Boolean isSystemModelFile() 
   + boolean isExternal() 
   + Object accept(Object,Object) 
   + ModelManager getModelManager() 
   + string[] getImports() 
   ~ void validate() throws IllegalModelException
   + string getImportedType(string) 
   + boolean isDefined(string) 
   + ClassDeclaration getLocalType(string) 
   + AssetDeclaration getAssetDeclaration(string) 
   + TransactionDeclaration getTransactionDeclaration(string) 
   + EventDeclaration getEventDeclaration(string) 
   + ParticipantDeclaration getParticipantDeclaration(string) 
   + string getNamespace() 
   + string getName() 
   + AssetDeclaration[] getAssetDeclarations() 
   + TransactionDeclaration[] getTransactionDeclarations() 
   + EventDeclaration[] getEventDeclarations() 
   + ParticipantDeclaration[] getParticipantDeclarations() 
   + ClassDeclaration[] getClassDeclarations() 
   + ConceptDeclaration[] getConceptDeclarations() 
   + EnumDeclaration[] getEnumDeclarations() 
   + MapDeclaration[] getMapDeclarations() 
   + ScalarDeclaration[] getScalarDeclarations() 
   + Object[] getDeclarations(Function) 
   + ClassDeclaration[] getAllDeclarations() 
   + string getDefinitions() 
   + object getAst() 
   + string getConcertoVersion() 
   + void isCompatibleVersion() 
   + void enforceImportVersioning(imp) 
}
class ParticipantDeclaration extends IdentifiedDeclaration {
   + void constructor(ModelFile,Object) throws IllegalModelException
   + string declarationKind() 
}
class Property extends Decorated {
   + void constructor(ClassDeclaration,Object) throws IllegalModelException
   + ModelFile getModelFile() 
   + ClassDeclaration getParent() 
   ~ void validate(ClassDeclaration) throws IllegalModelException
   + string getName() 
   + string getType() 
   + boolean isOptional() 
   + string getFullyQualifiedTypeName() 
   + string getFullyQualifiedName() 
   + string getNamespace() 
   + boolean isArray() 
   + boolean isTypeEnum() 
   + boolean isPrimitive() 
}
class RelationshipDeclaration extends Property {
   + void constructor(ClassDeclaration,Object) throws IllegalModelException
   ~ void validate(ClassDeclaration) throws IllegalModelException
   + String toString() 
   + boolean isRelationship() 
}
class ScalarDeclaration extends Declaration {
   ~ void validate() throws IllegalModelException
   + Boolean isIdentified() 
   + Boolean isSystemIdentified() 
   + string getIdentifierFieldName() 
   + string getType() 
   + string getSuperType() 
   + ClassDeclaration getSuperTypeDeclaration() 
   + Validator getValidator() 
   + void getDefaultValue() 
   + String toString() 
   + boolean isAbstract() 
   + boolean isScalarDeclaration() 
   + boolean isAsset() 
   + boolean isParticipant() 
   + boolean isTransaction() 
   + boolean isEvent() 
   + boolean isConcept() 
}
class TransactionDeclaration extends IdentifiedDeclaration {
   + void constructor(ModelFile,Object) throws IllegalModelException
   + string declarationKind() 
}
class MetamodelException extends BaseException {
   + void constructor(string) 
}
class Identifiable extends Typed {
   ~ void constructor(ModelManager,ClassDeclaration,string,string,string,string) 
   + string getTimestamp() 
   + string getIdentifier() 
   + void setIdentifier(string) 
   + string getFullyQualifiedIdentifier() 
   + String toString() 
   + boolean isRelationship() 
   + boolean isResource() 
   + String toURI() 
}
class Relationship extends Identifiable {
   + String toString() 
   + boolean isRelationship() 
   + Relationship fromURI(ModelManager,String,String?,String?) 
}
class Resource extends Identifiable {
   + String toString() 
   + boolean isResource() 
   + boolean isConcept() 
   + boolean isIdentifiable() 
   + Object toJSON() 
}
class Typed {
   ~ void constructor(ModelManager,ClassDeclaration,string,string) 
   + string getType() 
   + string getFullyQualifiedType() 
   + string getNamespace() 
   + void setPropertyValue(string,string) 
   + void addArrayValue(string,string) 
   + boolean instanceOf(String) 
   ~ void toJSON() 
}
class ValidatedResource extends Resource {
   + void setPropertyValue(string,string) throws Error
   + void addArrayValue(string,string) throws Error
   + void validate() throws Error
}
class ModelLoader {
   + ModelManager[] loadModelManager(string[],object,boolean?,boolean?,number?) 
   + ModelManager[] loadModelManagerFromModelFiles(object[],string[],object,boolean?,boolean?,number?) 
}
class ModelManager extends BaseModelManager {
   + void constructor(object?,boolean?,Object?,boolean?,boolean?) 
   + ModelFile addCTOModel(string,string?,boolean?) throws IllegalModelException
}
   + object getRootModel() 
class SecurityException extends BaseException {
   + void constructor(string) 
}
class Serializer {
   + void constructor(Factory,ModelManager,object?) 
   + void setDefaultOptions(Object) 
   + Object toJSON(Resource,Object?,boolean?,boolean?,boolean?,boolean?,boolean?,number?) throws Error
   + Resource fromJSON(Object,Object?,boolean,boolean,number?,boolean?) 
}
class TypeNotFoundException extends BaseException {
   + void constructor(string,string|,string,string) 
   + string getTypeName() 
}
