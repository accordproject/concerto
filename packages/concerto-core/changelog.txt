#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# This is the changelog for Hyperledger Concerto. It lists all significant changes to
# functionality and public API.
#
# The most recent entry in the changelog should be at the top of this file
# and must follow the format: Version <number> {public api digest} <date>. The version
# number must match that specified in package.json. The public api digest is computed
# and validated using ./scripts/api-changelog.sh
#
# Any changes to the public API must have an associated changelog entry.
#
# Note that the latest public API is documented using JSDocs and is available in api.txt.
#

Version 3.17.5 {9bd69f9522c14a99a085f077e12ac4b2} 2024-08-29
- importAliasing added to ModelManager parameters

Version 3.17.4 {8cb49c7092c48b568da3b0a3eeab2777} 2024-08-01
- Added api getImportedType() in modeFile

Version 3.17.3 {1c15de121dd80375cf531bc2244efdd0} 2024-07-26
- Added new methods to extract vocab or non-vocab decorators from model
- Fixed some minor bugs related to vocabulary parsing
- Exposed method to validate locale

Version 3.17.2 {eb3903401fdcf7c26cca1f3f8a029171} 2024-07-17
- Added new optimized methods for decorating models with decorator commandsets
- fixed other decorator command set bugs

Version 3.17.1 {ddc91ebd1ff660b421b60302d1e92271} 2024-06-21
- Added 'enableAliasedType' option to BaseModelManager
- Aliased types mapped to FQN in modelfile

Version 3.16.8 {a23d37a4a92071314ff821f879943364} 2024-07-09
- Added a new pathway for applying dcs target at namespace

Version 3.16.7 {8f455df1e788c4994f423d6e236bee21} 2024-05-01
- Added missing `strictQualifiedDateTimes` option to Serializer.fromJSON

Version 3.13.4 {56df3674b9e8d094cec0ae690c07ea1f} 2024-04-20
- Mark accept methods in introspect classes as public
- Generate changes in TypeScript definitions

Version 3.13.3 {8f59b43e6071c4d3ae42e94476142f7a} 2023-11-07
- Added DCS and vocabulary extraction support for decoratorManager
- Added errortype to BaseException and used that to define error types in introspect

Version 3.13.2 {dccc690753912cf87e7ceec56d949058} 2023-10-18
- Add getNamespace method to key type and value type of maps

Version 3.13.1 {f5a9a1ea6a64865843a3abb77798cbb0} 2023-10-18
- Add migrate option to DecoratorManager options

Version 3.13.1 {f435a20a00712e49c5cd32bc73ecb06a} 2023-10-03
- Add JSDoc for enableMapType option on ModelManager

Version 3.13.1 {6b09c1c58abcc77eecbb44e375c2efb8} 2023-10-03
- Add enableMapType option to BaseModelManager options

Version 3.13.0 {125b7f97f8740628b2629b2793384cc7} 2023-10-03
- Update DecoratorManager to support multiple value compare
- Create DecoratorManager.validate method to validate structure of decorator command set

Version 3.12.4 {7738d5490ea8438677e1d21d704bb5aa} 2023-08-31
- Adds validate and validateCommands options to DecoratorManager.decorateModels
- Adds addMetamodel option to BaseModelManager

Version 3.12.2 {bb8951ea059e312cd347b715bc9fc8d7} 2023-08-28
- Adds ModelFile :: getClassDeclarations

Version 3.10.1 {399b057614d1178c02b744184b09f845} 2023-08-22
- Change return type for MapDeclaration :: MapKeyType MapValueType

Version 3.10.0 {3bba55da8dbb522de132359415eb7eeb} 2023-08-11
- Add MetamodelException

Version 3.8.5 {9cd54e67c4a39a3041706e4b37dce9f1} 2023-08-14
- Makes MapKeyType MapValueType processType() private

Version 3.8.4 {a99786670d2561bf2b84cb3eb1bb1a08} 2023-08-14
- Move isValidMapKey & isValidMapValue to ModelUtils

Version 3.8.3 {7dd828f29c49cedfb8eee2f3c6fefa6e} 2023-08-10
- Removes private identifiers from MapKeyType & MapValueType

Version 3.8.2 {1a497711c4cca84d1dfa92c0b581fadf} 2023-08-10
- Add isValidMapKey(Object) isValidMapValue(Object)

Version 3.8.1 {794268f69b81f05f711d38a9ef1a7833} 2023-08-1
- Add To MapKeyType, MapValue functionality

Version 3.7.0 {a97cb6ebd45679354ba4da1940d2bb8d} 2023-05-19
- Add MapDeclaration, MapKeyType, AggregateValueType

Version 3.5.0 {b419dc1b3db9662d8c941a044cd8db5a} 2023-03-07
- Introduce Declaration type. ClassDeclaration and Scalar Declaration both extend Declaration.

Version 3.0.0 {5fac664420fea3649a7a304941f190f1} 2022-08-28
- Allow client-provided RegExp engine to ModelManager
- Allow decorators to be attached to model files/namespaces

Version 2.3.0 {774980a857090905fe276b6e94f1dbb1} 2022-07-21
- Versioned namespaces

Version 2.3.0 {28dd70fd922b3b8b491216661796b95b} 2022-06-21
- Fix TypeScript types for MetaModel
- Fix TypeScript types for ModelManager

Version 2.2.0 {3d3fc2a6b3cd623b807ca0c4fc53680d} 2022-06-01
- Fix TypeScript types for ModelLoader

Version 2.0.2 {c9e7b234a5786db232520137be0a0fa8} 22-05-25
- Add ClassDeclaration.getDirectSubclasses for use in visitor

Version 2.0.1 {f5d85447683b22d6f4072f9cd7f8986e} 22-05-13
- Add Decorator.isDecorator for use in visitor

Version 2.0.1 {62a8de0d35789bbe523269e4e29cd8e6} 2022-04-25
- Correct type for Concerto.getModelManager()

Version 2.0.0-alpha.1 {292b93fc879d3b0c3969711b24312ffd} 2022-03-30
- Remove custom instanceof and add methods to check runtime type
- Remove support for Node 12
- Generate Typescript definitions from JSDoc
- Parser directly constructs metamodel instance
- Convert MetaModel to a class for Typescript + Webpack compatability
- Update Acorn to latest and refactor the JavaScript parser so we can use static class members
- Add DecoratorManager
- New packages for parsing and metamodel, distinct from core
- Revisions to model manager API to support various inputs (e.g., CTO or AST)
- New vocabulary package for handling model vocabularies

Version 1.2.1 {b19318bb094e5da7bdff192cf9a3b4f2} 2021-08-12
- Fixes to metamodel, including terminology changes
- Ability to roundtrip model manager to metamodel
- Expose name resolution to MetaModel API

Version 1.1.2 {2c9512d9d90bde289b47942937d252ca} 2021-08-12
- Add Factory.newId for creating new unique IDs

Version 1.1.0 {16fb2d5684ec917532a19428c74f1ebf} 2021-07-13
- Add support for Concerto metamodel with import/export to CTO

Version 1.0.3 {1fe469fe1a79af5d5a4f5ec7dee6b7d4} 2021-06-25
- Aligns JSDoc and the TypeScript interface

Version 1.0.0-alpha.8 {08b0c809336354e7d1b17a2e44fca8b0} 2021-04-22
- Add datetimeutil & Consolidate UTC offset support
- Delete Concerto.instanceOf replace by ModelManager.derivesFrom

Version 1.0.0-alpha.7 {aec0f0d4e25be4dab03d6b94be60767b} 2021-04-17
- Remove unused systemtime options in modelfile API
- Fixes to the validation option targeting the ergo ES6 runtime

Version 1.0.0-alpha.5 {446997953e3ac90783ff9d492a8e7fdc} 2021-04-08
- Support Concerto version pragma at the beginning of model files
- Remove allowEmptyId option in factory API

Version 1.0.0-alpha.4 {3b7ebc06a536a5e81624225c7b0079ee} 2021-04-08
- Allow optional UTC offset for validating DateTime values

Version 1.0.0-alpha.3 {7b8763c243f937d84579350aef348c3c} 2020-12-25
- Concepts may now be identified
- assets/participants have a system identifier created ($identifier) automatically
- /transaction/events have a system timestamp created ($timestamp) automatically
- Can no longer redefine the identifier declared by a super type
- 'concerto' namespace is now reserved and cannot be used by user code
- All types extend concerto.Concept
- Removed System Models APIs
- Remove notion of system table
- Remove user ability to change system model
- Update JsDoc for ModelManager.addModelFiles
- Add Logger API
- Add instanceof alternatives to Factory and Serializer
- Add new Concerto API (experimental)

Version 0.82.9 {c96bafdd4a2bb7402f6b520a0cefe537} 2020-09-09
- Add option parameters to ModelLoader

Version 0.82.6 {03fa6481ffdf0e58cb110c6f24009d18} 2020-02-26
- Update JsDoc for ModelManager.addModelFiles

Version 0.82.1 {dee013e99a3c2d6acc4eddfb00aad2a2} 2019-10-22
- Make several constructors public
- Add model loader utility class

Version 0.80.3 {6f5a9ab45943cb76732c14b11f47d044} 2019-08-24
- Add Ergo option to serializer

Version 0.80.1 {297c88d29ce911ec6efc0f28ceeeb660} 2019-08-24
- Adds getModels and writeModelsToFileSystem functions to ModelManager
- Fixes API generation for hasSymbol function

Version 0.71.7 {2b5649f2b8f13d52241959f613cbfdd7} 2019-07-29
- Add static hasInstance as alternative instanceof implementation for AssetDeclaration, ClassDeclaration, EnumDeclaration, ConceptDeclaration, EnumValueDeclaration, Field, ModelFile, ModelManager, RelationshipDeclaration, TransactionDeclaration

Version 0.70.5 {db48b7eb8404d9206e9bc8efb3de0009} 2019-06-23
- Update exception triggered when creating EventDeclaration class

Version 0.31.0 {82c0c10648bd10fb79b84db3fcda1776} 2018-10-03
- Remove built-in system model file

Version 0.30.1 {fd4b813bc0bda4042db3c4657893593f} 2018-10-03
- Remove Wallet

Version 0.30.0 {785896a0c4c91e70dcbbfcc5bab33938} 2018-10-02
- Refactor to make model manager its own module
