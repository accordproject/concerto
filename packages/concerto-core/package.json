{
  "name": "@accordproject/concerto-core",
  "version": "3.19.1",
  "description": "Core Implementation for the Concerto Modeling Language",
  "homepage": "https://github.com/accordproject/concerto",
  "engines": {
    "node": ">=18",
    "npm": ">=10"
  },
  "main": "index.js",
  "browser": "dist/concerto-core.js",
  "typings": "types/index.d.ts",
  "scripts": {
    "prepublishOnly": "npm run webpack",
    "pretest": "npm run lint",
    "lint": "eslint .",
    "postlint": "npm run licchk",
    "licchk": "license-check-and-add",
    "postlicchk": "npm run doc",
    "doc": "jsdoc --pedantic --recurse -c jsdoc.json",
    "test": "node ./scripts/api-changelog.js && cross-env TZ=UTC nyc mocha --recursive -t 10000",
    "test:watch": "cross-env TZ=UTC nyc mocha --watch --recursive -t 10000",
    "mocha": "mocha --recursive -t 10000",
    "nyc": "nyc mocha --recursive -t 10000",
    "build": "npm run build:types",
    "postbuild": "npm run webpack",
    "webpack": "webpack --config webpack.config.js --mode production",
    "build:types": "tsc index.js --declaration --allowJs --emitDeclarationOnly --outDir types"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/accordproject/concerto.git",
    "directory": "packages/concerto-cto"
  },
  "keywords": [
    "schema",
    "data model",
    "verification"
  ],
  "author": "accordproject.org",
  "license": "Apache-2.0",
  "devDependencies": {
    "@babel/preset-env": "7.16.11",
    "acorn": "8.5.0",
    "acorn-walk": "8.2.0",
    "babel-loader": "8.2.3",
    "chai": "4.3.6",
    "chai-as-promised": "7.1.1",
    "chai-things": "0.2.0",
    "commander": "2.20.0",
    "cross-env": "7.0.3",
    "doctrine": "3.0.0",
    "eslint": "8.2.0",
    "fs-extra": "9.1.0",
    "jsdoc": "^4.0.2",
    "klaw": "3.0.0",
    "license-check-and-add": "2.3.6",
    "mocha": "10.0.0",
    "mockery": "2.1.0",
    "moxios": "0.4.0",
    "node-polyfill-webpack-plugin": "2.0.1",
    "nyc": "15.1.0",
    "process": "0.11.10",
    "sinon": "12.0.0",
    "sinon-chai": "3.7.0",
    "tmp-promise": "3.0.2",
    "typescript": "4.6.3",
    "webpack": "5.94.0",
    "webpack-cli": "4.9.1",
    "xregexp": "5.1.1",
    "yargs": "17.3.1"
  },
  "dependencies": {
    "@accordproject/concerto-cto": "3.19.1",
    "@accordproject/concerto-metamodel": "3.10.1",
    "@accordproject/concerto-util": "3.19.1",
    "dayjs": "1.11.10",
    "debug": "4.3.4",
    "lorem-ipsum": "2.0.8",
    "randexp": "0.5.3",
    "semver": "7.5.4",
    "slash": "3.0.0",
    "urijs": "1.19.11",
    "uuid": "9.0.1"
  },
  "browserslist": "> 0.25%, not dead",
  "license-check-and-add-config": {
    "folder": "./lib",
    "license": "HEADER",
    "exact_paths_method": "EXCLUDE",
    "exact_paths": [
      "api.txt",
      "composer-logs",
      "coverage",
      "index.d.ts",
      "./system",
      "LICENSE",
      "node_modules",
      ".nyc-output",
      "out",
      "dist",
      ".tern-project"
    ],
    "file_type_method": "EXCLUDE",
    "file_types": [
      ".yml",
      ".yaml",
      ".zip",
      ".tgz"
    ],
    "insert_license": false,
    "license_formats": {
      "js|njk|pegjs|cto|acl|qry": {
        "prepend": "/*",
        "append": " */",
        "eachLine": {
          "prepend": " * "
        }
      },
      "npmrc|editorconfig|txt": {
        "eachLine": {
          "prepend": "# "
        }
      },
      "md": {
        "file": "HEADER.md"
      }
    }
  },
  "nyc": {
    "produce-source-map": "true",
    "sourceMap": "inline",
    "reporter": [
      "lcov",
      "text-summary",
      "html",
      "json"
    ],
    "include": [
      "lib/**/*.js"
    ],
    "exclude": [],
    "all": true,
    "check-coverage": true,
    "statements": 99,
    "branches": 97,
    "functions": 98,
    "lines": 99
  }
}
