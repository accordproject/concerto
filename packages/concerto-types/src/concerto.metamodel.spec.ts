/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ModelManager, ModelFile, Concerto } from '@accordproject/concerto-core';
const { MetaModelUtil, MetaModelNamespace } = require('@accordproject/concerto-metamodel');
import { createGenerator } from "ts-json-schema-generator";
import Ajv from "ajv";
import path from "path";

test('Chained TypeScript and JSONSchema conversion respects inheritance when flattening subclasses to a union type', () => {

    // TS to JSON Schema conversion
    const config = {
        path: path.resolve('../src/generated/union/concerto-metamodel@1.0.0.ts'),
        tsconfig: path.resolve('../concerto-types/tsconfig.json'),
        type: 'IDecorator',
    }
    const jsonSchema = createGenerator(config).createSchema(config.type);
    expect(jsonSchema).toMatchSnapshot();

    // Test instance
    const data = {
        $class: `concerto.metamodel@1.0.0.Decorator`,
        name: 'displayName',
        arguments: [
            {
                value: 'Account ID',
                $class: `concerto.metamodel@1.0.0.DecoratorString`,
            }
        ],
    }

    // Validate the instance with Ajv, a JSON Schema validator
    const ajv = new Ajv()
    const validate = ajv.compile(jsonSchema)
    const valid = validate(data)
    expect(validate.errors).toBeNull();

    // Validate the instance with Concerto
    const modelManager = new ModelManager();
    modelManager.addModelFile(new ModelFile(modelManager, MetaModelUtil.metaModelAst, undefined, MetaModelNamespace));
    const concerto = new Concerto(modelManager);
    expect(() => concerto.validate(data)).not.toThrow();
});
