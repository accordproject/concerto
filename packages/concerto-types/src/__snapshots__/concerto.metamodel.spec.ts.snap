// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chained TypeScript and JSONSchema conversion respects inheritance when flattening subclasses to a union type 1`] = `
{
  "$ref": "#/definitions/IDecorator",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "definitions": {
    "DecoratorLiteralUnion": {
      "anyOf": [
        {
          "$ref": "#/definitions/IDecoratorString",
        },
        {
          "$ref": "#/definitions/IDecoratorNumber",
        },
        {
          "$ref": "#/definitions/IDecoratorBoolean",
        },
        {
          "$ref": "#/definitions/IDecoratorTypeReference",
        },
      ],
    },
    "IDecorator": {
      "additionalProperties": false,
      "properties": {
        "$class": {
          "type": "string",
        },
        "arguments": {
          "items": {
            "$ref": "#/definitions/DecoratorLiteralUnion",
          },
          "type": "array",
        },
        "location": {
          "$ref": "#/definitions/IRange",
        },
        "name": {
          "type": "string",
        },
      },
      "required": [
        "$class",
        "name",
      ],
      "type": "object",
    },
    "IDecoratorBoolean": {
      "additionalProperties": false,
      "properties": {
        "$class": {
          "type": "string",
        },
        "location": {
          "$ref": "#/definitions/IRange",
        },
        "value": {
          "type": "boolean",
        },
      },
      "required": [
        "$class",
        "value",
      ],
      "type": "object",
    },
    "IDecoratorNumber": {
      "additionalProperties": false,
      "properties": {
        "$class": {
          "type": "string",
        },
        "location": {
          "$ref": "#/definitions/IRange",
        },
        "value": {
          "type": "number",
        },
      },
      "required": [
        "$class",
        "value",
      ],
      "type": "object",
    },
    "IDecoratorString": {
      "additionalProperties": false,
      "properties": {
        "$class": {
          "type": "string",
        },
        "location": {
          "$ref": "#/definitions/IRange",
        },
        "value": {
          "type": "string",
        },
      },
      "required": [
        "$class",
        "value",
      ],
      "type": "object",
    },
    "IDecoratorTypeReference": {
      "additionalProperties": false,
      "properties": {
        "$class": {
          "type": "string",
        },
        "isArray": {
          "type": "boolean",
        },
        "location": {
          "$ref": "#/definitions/IRange",
        },
        "type": {
          "$ref": "#/definitions/ITypeIdentifier",
        },
      },
      "required": [
        "$class",
        "isArray",
        "type",
      ],
      "type": "object",
    },
    "IPosition": {
      "additionalProperties": false,
      "properties": {
        "$class": {
          "type": "string",
        },
        "column": {
          "type": "number",
        },
        "line": {
          "type": "number",
        },
        "offset": {
          "type": "number",
        },
      },
      "required": [
        "$class",
        "column",
        "line",
        "offset",
      ],
      "type": "object",
    },
    "IRange": {
      "additionalProperties": false,
      "properties": {
        "$class": {
          "type": "string",
        },
        "end": {
          "$ref": "#/definitions/IPosition",
        },
        "source": {
          "type": "string",
        },
        "start": {
          "$ref": "#/definitions/IPosition",
        },
      },
      "required": [
        "$class",
        "end",
        "start",
      ],
      "type": "object",
    },
    "ITypeIdentifier": {
      "additionalProperties": false,
      "properties": {
        "$class": {
          "type": "string",
        },
        "name": {
          "type": "string",
        },
        "namespace": {
          "type": "string",
        },
      },
      "required": [
        "$class",
        "name",
      ],
      "type": "object",
    },
  },
}
`;
